keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,423 @@
++--------+----------+--------+----------+--------+--------+--------+
+|   pod  | pod, pod |        | pod, pod |        |        |   pod  |
++--------+----------+--------+----------+--------+--------+--------+
+```
+
+Now let's say there comes a pod, it wants to be placed along with pods which
+carries label `foo` in zones.
+
+If global topologyKey is "zone" and maxSkew is "1", then incoming pod can only
+be put into zone3 because for zone1, it violates `matching num (3) - min
+matching num (1) < maxSkey (1)`. Zone2 violate the formula the same way.
+
+If global topologyKey is "node" and masSkew is "1", things are slightly
+different. Min matching num becomes 0 now, and hence only node1c, node2b and
+node2c are qualified candidates.
+
+- what if a topology domain is infeasible
+
+Suppose we have pods distribution in a 3-zone cluster as 3/3/0, and all pods
+have label `foo`:
+
+```
++-------------+-------------+--------------------+
+|    zone1    |    zone2    | zone3 (infeasible) |
++-------------+-------------+--------------------+
+| pod,pod,pod | pod,pod,pod |                    |
++-------------+-------------+--------------------+
+```
+
+And we have an incoming pod which wants to be scheduled with pods which carry
+label `foo` in zones. And suppose all nodes in zone3 are infeasible, e.g. due to
+taints or lack of resources. In this case:
+
+If it's a hard requirement, we treat the `min matching num` as 0, which means
+incoming pod would fail to be scheduled.
+
+If it's a soft requirement, we treat the `min matching num` as 3 instead of 0,
+which means incoming pod can be placed onto zone1 or zone2.
+
+- (more cases) when a topoloy domain is infeasible
+
+    > Suppose maxSkew is 1: (~~zone~~ means the zone is infeasible)
+
+    - for a "1/1/~~0~~" cluster, pod can't be placed onto any zone if it's a
+      Predicate; zone1 and zone2 are equally preferred if it's a Priority
+    - for a "2/1/~~0~~" cluster, pod can't be placed onto any zone if it's a
+      Predicate; zone2 is preferable over zone1 if it's a Priority
+    - for a "1/1/~~1~~" cluster, pod can be placed onto zone1 or zone2 if it's a
+      Predicate; zone1 and zone2 are equally preferred if it's a Priority
+    - for a "2/1/~~1~~" cluster, pod can be placed onto zone2 if it's a
+      Predicate; zone2 is preferable over zone1 if it's a Priority
+
+- when formula check is enforced
+
+We only enforce the formula check upon new pod scheduling. In other words, if
+pods become imbalanced (due to explicit taints, lack of resources, or node
+lost), we don't do proactive re-scheduling. Our goal is to not make things
+worse.
+
+### How User Stories are Addressed
+
+Both story 1 and story 2 can be addressed using `TopologySpreadConstraint`.
+
+And the pseudo algorithms below explain the processing flow in a nutshell.
+
+- Predicate
+
+```bash
+for each candidate node; do
+    if "TopologySpreadConstraint" is enabled; then
```

Comment:
nit: `is enabled for the pod being scheduled` ?

Reply:
Done.